From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: KurumiFake <kurumiisshidowife@gmail.com>
Date: Sun, 26 Dec 2021 18:13:24 +0700
Subject: [PATCH] lithium: cache chunk gen sea level


diff --git a/src/main/java/net/minecraft/world/level/levelgen/ChunkGeneratorAbstract.java b/src/main/java/net/minecraft/world/level/levelgen/ChunkGeneratorAbstract.java
index 916e893cf771a2ea825d60104115282295529119..f39b0b2b259cf9cecf460331b789bd8009090677 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/ChunkGeneratorAbstract.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/ChunkGeneratorAbstract.java
@@ -52,6 +52,7 @@ import net.minecraft.world.level.levelgen.synth.NoiseGeneratorPerlin;
 
 public final class ChunkGeneratorAbstract extends ChunkGenerator {
 
+    private int cachedSeaLevel; // Yatopia - lithium cache chunk gen settings
     public static final Codec<ChunkGeneratorAbstract> d = RecordCodecBuilder.create((instance) -> {
         return instance.group(WorldChunkManager.a.fieldOf("biome_source").forGetter((chunkgeneratorabstract) -> {
             return chunkgeneratorabstract.b;
@@ -136,6 +137,7 @@ public final class ChunkGeneratorAbstract extends ChunkGenerator {
         } else {
             this.v = null;
         }
+        this.cachedSeaLevel = ((GeneratorSettingBase) this.h.get()).g(); // Yatopia - lithium cache chunk gen settings
 
     }
 
@@ -723,10 +725,18 @@ public final class ChunkGeneratorAbstract extends ChunkGenerator {
         return this.x;
     }
 
+    // Yatopia start - lithium cache chunk gen settings
+    /**
+     * Use cached sea level instead of retrieving from the registry every time.
+     * This method is called for every block in the chunk so this will save a lot of registry lookups.
+     *
+     * @author SuperCoder79
+     */
     @Override
-    public int getSeaLevel() {
-        return ((GeneratorSettingBase) this.h.get()).g();
+    public int getSeaLevel() { // Cached method of getting sealevel
+        return this.cachedSeaLevel;
     }
+    // Yatopia end
 
     @Override
     public List<BiomeSettingsMobs.c> getMobsFor(BiomeBase biomebase, StructureManager structuremanager, EnumCreatureType enumcreaturetype, BlockPosition blockposition) {
