From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: KurumiFake <kurumiisshidowife@gmail.com>
Date: Thu, 30 Dec 2021 17:25:52 +0000
Subject: [PATCH] rm


diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 8906354a26d602c9c9216e5b904d322c8c36f851..4fd45f8fbf9fcbf5c3613dd42daa3bf66ff4454f 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -248,12 +248,6 @@ public class Main {
                 return;
             }
 
-            float javaVersion = Float.parseFloat(System.getProperty("java.class.version"));
-            if (javaVersion > 60.0) {
-                System.err.println("Unsupported Java detected (" + javaVersion + "). Only up to Java 16 is supported.");
-                if (!Boolean.getBoolean("Paper.IgnoreJavaVersion")) return; // Paper
-            }
-
             try {
                 // Paper start - Handled by TerminalConsoleAppender
                 /*
